protocols/autonat/src/protocol.rs

```diff
@@ -269,35 +270,38 @@ impl DialResponse {
     }
 
     pub fn into_bytes(self) -> Vec<u8> {
+        use quick_protobuf::MessageWrite;
+
         let dial_response = match self.result {
-            Ok(addr) => structs_proto::message::DialResponse {
-                status: Some(0),
-                status_text: self.status_text,
+            Ok(addr) => proto::DialResponse {
+                status: Some(proto::ResponseStatus::OK),
+                statusText: self.status_text,
                 addr: Some(addr.to_vec()),
             },
-            Err(error) => structs_proto::message::DialResponse {
+            Err(error) => proto::DialResponse {
                 status: Some(error.into()),
-                status_text: self.status_text,
+                statusText: self.status_text,
                 addr: None,
             },
         };
 
-        let msg = structs_proto::Message {
-            r#type: Some(structs_proto::message::MessageType::DialResponse as _),
+        let msg = proto::Message {
+            type_pb: Some(proto::MessageType::DIAL_RESPONSE),
             dial: None,
-            dial_response: Some(dial_response),
+            dialResponse: Some(dial_response),
         };
 
-        let mut bytes = Vec::with_capacity(msg.encoded_len());
-        msg.encode(&mut bytes)
-            .expect("Vec<u8> provides capacity as needed");
-        bytes
+        let mut buf = Vec::with_capacity(msg.get_size());
+        let mut writer = Writer::new(&mut buf);
+        msg.write_message(&mut writer).expect("Encoding to succeed");
```

Comment:
You are completely on point. It think it would be useful to upstream the following to `quick-protobuf`:


```rust
/// A trait to handle deserialization based on parsed `Field`s
pub trait MessageWrite: Sized {

	// Other functions as currently defined ...

	fn write_to_vec(&self) -> Vec<u8> {
		let mut vec = Vec::with_capacity(self.get_size());
		let mut writer = Writer::new(&mut vec);

		self.write_message(&mut writer).expect("Vec<u8> provides enough capacity");

		vec
	}
}
```

On this note, I think there is some clean-up to be done in terms of APIs. It is quite confusing that there is `MessageWrite::write_message` which writes to a `Writer` and there is also `Writer::write_message` which writes the message using `MessageWriter::write_message` **with a length-prefix**!

Reply:
Opened an issue here: https://github.com/tafia/quick-protobuf/issues/246

Reply:
> Because we only expect errors on memory allocation and given that we are using a Vec which can dynamically allocate, this is not an issue.

It wasn't clear from the documentation and code in what cases we should expect errors https://github.com/libp2p/rust-libp2p/pull/3066#discussion_r1006239290 so that is why the message is so vague. Do we want this changed in this PR or for changes upstream?

Reply:
I am fine merging as is.